### PR TITLE
Remove GetRepositoryByRef and add GetRepositoryByOwnerAndName

### PR DESCRIPTION
**Pull request by lunny**
This PR changed many places from 2 database query to 1 database query and simple the codes.

### CODE REVIEW

File: routers/repo/http.go [hunk 1]
```diff
@@ -64,22 +64,12 @@ func HTTP(ctx *context.Context) {
...
-		}
-		return
-	}
-
-	repo, err := models.GetRepositoryByName(repoUser.ID, reponame)
+	repo, err := models.GetRepositoryByOwnerAndName(username, reponame)
 	if err != nil {
 		if models.IsErrRepoNotExist(err) {
-			ctx.Handle(http.StatusNotFound, "GetRepositoryByName", nil)
+			ctx.Handle(http.StatusNotFound, "GetRepositoryByOwnerAndName", nil)
```
Comment (ethantkoenig) [inline routers/repo/http.go]: nit: Can replace the `if`/`else` with `ctx.NotFoundOrServerError("GetRepositoryByOwnerAndName", models.IsErrRepoNotExist, err)`.

File: models/repo.go [hunk 2]
```diff
@@ -1891,21 +1891,20 @@ func DeleteRepository(doer *User, uid, repoID int64) error {
...
+	var repo Repository
+	has, err := x.Select("repository.*").
+		Join("INNER", "user", "`user`.id = repository.owner_id").
+		Where("repository.lower_name = ?", strings.ToLower(repoName)).
+		And("`user`.lower_name = ?", strings.ToLower(ownerName)).
+		Get(&repo)
 	if err != nil {
 		return nil, err
+	} else if !has {
+		return nil, ErrRepoNotExist{0, 0, ownerName + "/" + repoName}
```
Comment (ethantkoenig) [inline models/repo.go]: nit: Elsewhere in the code, the `Name` field of the `ErrRepoNotExist` struct is just the repo name (e.g. in `GetRepositoryByName`), but here it is both the owner's name and repo name. Could we instead add a `OwnerName` field, and keep the `Name` field only for the repo name?

File: models/issue.go [hunk 3]
```diff
@@ -974,7 +974,12 @@ func GetIssueByRef(ref string) (*Issue, error) {
 		return nil, errInvalidIssueNumber
 	}
 
-	repo, err := GetRepositoryByRef(ref[:n])
+	i := strings.IndexByte(ref[:n], byte('/'))
```
Comment (ethantkoenig) [inline models/issue.go]: nit: remove unnecessary cast `i := strings.IndexByte(ref[:n], '/')`

**Review by ethantkoenig — COMMENTED** [review]
A few nits

**Comment by lafriks** [comment]
LGTM

**Comment by lunny** [comment]
@ethantkoenig all done.

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Review by appleboy — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**